### DISCUSSION TITLE
Making my first mod for osu!

### DISCUSSION THREAD

**Discussion by osnesone** [ppy/osu#21258] [Q&A] [answered]
Hi, I am trying to make a mod for osu! named First Person Osu!. As the name implies it'll be a mod which will move the HitObjects while keeping the Cursor still to try and replicate the feel of an FPS game.

Although I am new in C# I have some experience with C/C++ from porting an old mod of [Teeworlds](https://www.teeworlds.com) called [FNG](https://github.com/Jupeyy/teeworlds-fng2-mod/tree/fng_06) to the [DDNet](https://github.com/ddnet/ddnet) codebase and renamed it to [DFNG](https://github.com/Inateblig/dfng), so I thought I could take this challenge on. Turns out I need help :sweat_smile:

Here's how it I think it should work. At the start of the Update() it should get the Cursor's position and update the `HitObjects` accordingly. Then at the end of the `Update()` it should set the Cursor's position to the center of the screen. That way info from the mouse can still be retrieved while giving the illusion that the Cursor doesn't move.

For starters I copied the `OsuModMagnetised.cs` (under `osu/osu.Game.Rulesets.Osu/Mods`) to `OsuModFpOsu.cs` under the same folder. After that I changed the class name to `OsuModFpOsu`, and some other stuff to represent the mod.
Now I'm trying to have the cursor positioned at the center while retrieving its movement. The problem I am facing is that I don't know which property of `playfield.Cursor.AsNonNull().ActiveCursor` to change so it ends up with different mouse position and different cursor sprite rendering ![OsuBrokenCursor](https://user-images.githubusercontent.com/108964857/202283331-7b1053a6-b311-4daf-bbeb-4d41c1c4960c.png)

**Comment by peppy** [comment]
I don't think your approach is going to work. Hit objects should not be moved - only the playfield (aka viewport) should be moved.

**Reply by osnesone** [reply]
Ok, so now I've been trying to move the playfield so that the cursor is positioned at the center of the screen but it seems that the cursor is moved wrongly, as if it gets affected by the playfield movement. This is the Update() code.
```C#
public void Update(Playfield playfield)
{
    var curpos = playfield.Cursor.ActiveCursor.Position;

    Console.WriteLine("cp: {0}, plfp: {1}", curpos, playfield.Position);

    Vector2 basesize = Osu.UI.OsuPlayfield.BASE_SIZE;
    Vector2 newpos = (basesize / 2 - curpos) * MouseSensitivity.Value;

    if (Math.Abs(newpos.X) <= basesize.X && Math.Abs(newpos.Y) <= basesize.Y) {
        playfield.Position = newpos;
    }
}
```

**Reply by frenzibyte** [reply]
You will have to find a way to make `OsuInputManager` receive mouse event from parent and convert that into moving the playfield rather than passing it down as mouse input for its hierarchy (see `PassThroughInputManager`, which `OsuInputManager` inherits from and is the one generating input from events).

**Reply by osnesone** [reply]
I managed to get the mouse position in window coordinates using this:
```C#
private OsuInputManager inputManager = null!;

public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
{
    inputManager = (OsuInputManager)drawableRuleset.KeyBindingInputManager;
}

public void Update(Playfield playfield)
{
    var curpos = inputManager.CurrentState.Mouse.Position;

    Console.WriteLine("cp: {0}, plfp: {1}", curpos, playfield.Position);

    Vector2 basesize = Osu.UI.OsuPlayfield.BASE_SIZE;
    Vector2 newpos = (basesize / 2 - curpos) * MouseSensitivity.Value;

    if (Math.Abs(newpos.X) <= basesize.X && Math.Abs(newpos.Y) <= basesize.Y) {
        playfield.Position = newpos;
    }
}
```
I can't find a way to convert them to playfield coordinates. I spoted a function named `ToLocalSpace()` which is used in `OsuModBlinds.cs`.

**Reply by frenzibyte** [reply]
As I've mentioned, take a look at how input flow works in `InputManager`s (mainly how input gets passed from one manager to another, via `PassThroughInputManager`) and use that to intercept mouse input in `OsuInputManager` and move the playfield rather than moving mouse.

**Comment by frenzibyte** [comment, marked as answer]
Will mention that this was requested long ago and tracked in https://github.com/ppy/osu/discussions/13053.

**Reply by frenzibyte** [reply]
Move further questions you have in the discussion thread above.
